### PR TITLE
Added Phase Change Material constitutive class and heat conduction model

### PR DESCRIPTION
**Pull request by markleader**

### CODE REVIEW

File: tacs/constitutive.pyx [hunk 1]
```diff
@@ -465,6 +465,62 @@ cdef class PlaneStressConstitutive(Constitutive):
             self.ptr = NULL
             self.cptr = NULL
 
+cdef class PhaseChangeMaterialConstitutive(Constitutive):
+    """
+    This is the base class for the plane stress constitutive objects.
```
Comment (timryanb) [inline tacs/constitutive.pyx:470]: Can you update the doctoring here? It still describes this as a plane stress constitutive

File: src/constitutive/TACSConstitutive.h [hunk 2]
```diff
@@ -134,6 +139,30 @@ class TACSConstitutive : public TACSObject {
                                 const double pt[], const TacsScalar X[],
                                 int dvLen, TacsScalar dfdx[]) {}
 
+  virtual void addDensityDVSens(int elemIndex, TacsScalar scale,
```
Comment (timryanb) [inline src/constitutive/TACSConstitutive.h:142]: Looks like the docstring is missing here

Comment (timryanb) [reply src/constitutive/TACSConstitutive.h:142]: Missing docstring here too.

File: src/constitutive/TACSConstitutive.h [hunk 3]
```diff
@@ -134,6 +139,30 @@ class TACSConstitutive : public TACSObject {
...
+                                const double pt[], const TacsScalar X[],
+                                int dvLen, TacsScalar dfdx[],
+                                const TacsScalar u[]) {
```
Comment (timryanb) [inline src/constitutive/TACSConstitutive.h:145]: We might want to change the state variable name from `u` to `vars` to be consistent with the rest of TACS?

File: src/constitutive/TACSPhaseChangeMaterialConstitutive.cpp [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+    liquid_properties->incref();
+  }
+  lh = _lh;
+  Tm = _Tm;
+  t = _t;
+  tNum = _tNum;
+  tlb = _tlb;
+  tub = _tub;
+  dT = _dT;
+  b = 2.0 * tan(0.99 * M_PI / 2.0);
```
Comment (timryanb) [inline src/constitutive/TACSPhaseChangeMaterialConstitutive.cpp:50]: Can we add some comments to the code explaining how the formulation works? What the arctan is used for, each of the terms in the equation, etc.?

File: src/elements/TACSHeatConduction.cpp [hunk 5]
```diff
@@ -665,7 +665,7 @@ void TACSHeatConduction3D::getOutputData(
     int elemIndex, const double time, ElementType etype, int write_flag,
     const double pt[], const TacsScalar X[], const TacsScalar Ut[],
     const TacsScalar Ux[], int ld_data, TacsScalar *data) {
-  if (etype == TACS_SCALAR_2D_ELEMENT) {
+  if (etype == TACS_SCALAR_3D_ELEMENT) {
```
Comment (timryanb) [inline src/elements/TACSHeatConduction.cpp:668]: It actually looks like should be a `TACS_SOLID_ELEMENT` according to [here](https://github.com/smdogroup/tacs/blob/master/src/elements/TACSElement3D.cpp#L76).

@markleader I think we should also change [this](https://github.com/smdogroup/tacs/blob/master/src/elements/TACSHeatConduction.cpp#L308) to `TACS_PLANE_STRESS_ELEMENT`

@gjkennedy correct me if I'm wrong

File: tests/integration_tests/test_pcm_element_2d.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+bdf_file = os.path.join(base_dir, "./input_files/unit_plate.bdf")
+
+FUNC_REFS = {
+    "steady_state_avg_temp": 150.00000000000063,
+    "steady_state_ks_temp": 198.51811570667203,
+    "steady_state_mass": 0.9999999999999951,
+    "transient_avg_temp": 741.0557806913429,
+    "transient_ks_temp": 196.22817693313027,
+    "transient_mass": 4.999999999999405,
+}
```
Comment (timryanb) [inline tests/integration_tests/test_pcm_element_2d.py:24]: You can get rid of this FUNC_REF

File: tests/integration_tests/test_pcm_element_2d.py [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+from tacs import pytacs, TACS, elements, constitutive, functions, problems
+from pytacs_analysis_base_test import PyTACSTestCase
+
+"""
+Test heat conduction of a problem with non-uniform boundary conditions with initial conditions everywhere else.
+This is a unit-square domain with dT=100 on the left edge, dT=200 on the right edge, and initial condition of dT=150 in between.
+
+This example is a replication of the example here:
+https://kitchingroup.cheme.cmu.edu/blog/2013/03/07/Transient-heat-conduction-partial-differential-equations/
+"""
```
Comment (timryanb) [inline tests/integration_tests/test_pcm_element_2d.py:12]: Maybe mention somewhere in the docstring that version tests the PCM variant?

**Comment by timryanb** [comment]
Other than that, It looks good on my end. Thanks for the addition @markleader!

Check out steps 4 and 5 [here](https://github.com/smdogroup/tacs#contributing-to-tacs-development) for how to fix the black/clang linting failures

**Review by timryanb — APPROVED** [review]
Looks good to me. Let's give @gjkennedy a chance to review before finalizing

**Review by gjkennedy — APPROVED** [review]
